proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,301 @@
+
+## Metadata provider code-gen
+A run-time only [prototype](https://github.com/steveharter/runtime/tree/ExtConverters) which does not include the actual code generator was created by @steveharter (see sample generated code below). It makes public the existing internal metadata classes including `JsonClassInfo` and `JsonPropertyInfo`.
+
+It allows for minimal code gen and overlaps with other requested features including property ordering, before and after callbacks, and programmatic reading and writing of metadata.
+
+<details>
+  <summary>Snippets from prototype (click to expand)</summary>
+
+```cs
+public class WeatherForecast
+{
+    // The normal POCO properties:
+    public DateTime Date { get; set; }
+    public int TemperatureC { get; set; }
+    public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
+    public string Summary { get; set; }
+
+    // Static constructor code-generated; on first use of WeatherForcecast Type it is automatically registered.
+    static WeatherForecast()
+    {
+        // We need a way to get the correct options class; assume a global Serializer.Options for prototype (can vary per application).
+        JsonSerializerOptions options = Serializer.Options;
+
+        ClassInfo classInfo = new ClassInfo(options);
+        classInfo.Initialize();
+        options.Classes.TryAdd(typeof(WeatherForecast), classInfo);
+    }
+
+    // A nested code-generated private class to expose metadadata and handle get and set.
+    private class ClassInfo : JsonClassInfo
+    {
+        public ClassInfo(JsonSerializerOptions options) : base(typeof(WeatherForecast), options)
+        {
+            CreateObject = () => { return new WeatherForecast(); };
+        }
+
+        protected override IList<JsonPropertyInfo> GetProperties()
+        {
+            var p = new JsonPropertyInfo<DateTime>();
+            p.Options = Options;
+            p.Converter = Serializer.DateTimeConverter;
+            p.NameAsString = nameof(WeatherForecast.Date);
+
+            var properties = new List<JsonPropertyInfo>(4);
+
+            properties.Add(new JsonPropertyInfo<DateTime>
+            {
+                Options = Options,
+                Converter = Serializer.DateTimeConverter,
+                NameAsString = nameof(WeatherForecast.Date),
+                HasGetter = true,
+                HasSetter = true,
+                ShouldSerialize = true,
+                ShouldDeserialize = true,
+                // These delegates are nice that they work with aggressive linker
```

Comment:
How can the reflection generated good be ever faster than this? I think that this is the fastest you can get. I do not see what kind of magic can beat that.

Reply:
I was thinking if we have a fast way to call the getter\setter directly such as what you can do with MethodInfo.CreateDelegate but without having to use reflection to get the MethodInfo for a property.

If we could do that, it would be 2x as fast as the delegate "hop" approach as just as fast as IL Emit.

Here's pretty much every way to call a property getter (except for expressions):
```
Direct:2.
Reflection:831.
Manual loosely typed object delegate:49. **(what I have in prototype)**
Manual strongly typed delegate:28.
Generated static method which calls property:43.
MethodInfo getter:18. **(would be nice to be able to do something like this)**
Il getter:28.
Expression getter:11.
```

```cs
using System;
using System.Diagnostics;
using System.Linq.Expressions;
using System.Reflection;
using System.Reflection.Emit;

namespace ConsoleApp20benchmark
{
    class Program
    {
        static void Main(string[] args)
        {
            const long Iterations = 10_000_000;

            var poco = new POCO();
            PropertyInfo pi = typeof(POCO).GetProperty("MyProperty");
            MethodInfo mi = pi.GetGetMethod();

            int value;

            var sw = new Stopwatch();

            {
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = poco.MyProperty;
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Direct:{sw.ElapsedMilliseconds}.");
            }

            {
                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = (int)mi.Invoke(poco, null);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Reflection:{sw.ElapsedMilliseconds}.");
            }

            {
                Func<object, int> ManualGetter = (obj) =>
                {
                    return ((POCO)obj).MyProperty;
                };

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = ManualGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Manual loosely typed object delegate:{sw.ElapsedMilliseconds}.");
            }

            {
                Func<POCO, int> ManualStronglyTypedGetter = (obj) =>
                {
                    return obj.MyProperty;
                };

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = ManualStronglyTypedGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Manual strongly typed delegate:{sw.ElapsedMilliseconds}.");
            }

            {
                Func<POCO, int> ManualStronglyTypedGetter = (obj) => POCO.CallMyPropertyGetter(obj);

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = ManualStronglyTypedGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Generated static method which calls property:{sw.ElapsedMilliseconds}.");
            }

            {
                // Unforunately, we must use MethodInfo to get the direct delegate.
                Func<POCO, int> DelegateGetter = (Func<POCO, int>)mi.CreateDelegate(typeof(Func<POCO, int>));

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = DelegateGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"MethodInfo getter:{sw.ElapsedMilliseconds}.");
            }

            {
                var dynamicMethod = new DynamicMethod(
                    mi.Name,
                    mi.ReturnType,
                    new[] { typeof(object) },
                    typeof(Program).Module,
                    skipVisibility: true);

                ILGenerator generator = dynamicMethod.GetILGenerator();
                generator.Emit(OpCodes.Ldarg_0);
                generator.Emit(OpCodes.Castclass, mi.DeclaringType); // to verify type
                generator.Emit(OpCodes.Callvirt, mi);
                generator.Emit(OpCodes.Ret);

                Func<POCO, int> IlGetter = (Func<POCO, int>)dynamicMethod.CreateDelegate(typeof(Func<,>).MakeGenericType(mi.DeclaringType, mi.ReturnType));

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = IlGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();
                
                Console.WriteLine($"Il getter:{sw.ElapsedMilliseconds}.");
            }

            {
                ParameterExpression parameter = Expression.Parameter(pi.DeclaringType, pi.Name);
                MemberExpression property = Expression.Property(parameter, pi);
                Func<POCO, int> ExpressionGetter = (Func<POCO, int>) Expression.Lambda(property, parameter).Compile();

                sw.Reset();
                sw.Start();
                for (long l = 0; l < Iterations; l++)
                {
                    value = ExpressionGetter(poco);
                    Debug.Assert(value == 42);
                }
                sw.Stop();

                Console.WriteLine($"Expression getter:{sw.ElapsedMilliseconds}.");
            }
        }


        class POCO
        {
            public int MyProperty => 42;

            internal static int CallMyPropertyGetter(POCO obj)
            {
                return obj.MyProperty;
            }
        }
    }
}
```
```

Reply:
> Here's pretty much every way to call a property getter (except for expressions):

These numbers look suspect. I think you are seeing a lot of noise from tiered JIT, and other first-time initializations. I have added a big for-loop over the whole Main method. Here is the difference between 1st and 10th iteration:

1st iteration:
```
Direct:3.
Reflection:1233.
Manual loosely typed object delegate:80.
Manual strongly typed delegate:52.
Generated static method which calls property:80.
MethodInfo getter:37.
Il getter:44.
Expression getter:24.
```

10th iteration:
```
Direct:3.
Reflection:1104.
Manual loosely typed object delegate:26.
Manual strongly typed delegate:23.
Generated static method which calls property:26.
MethodInfo getter:30.
Il getter:33.
Expression getter:20.
```

Reply:
The 10th iteration numbers say that direct call is super fast (it should actually get optimized out completely in your benchmark and so you are just measuring how fast one can count to 10_000_000), the reflection Invoke is slow, and the rest is pretty much the same - the differences are in the noise range.

Reply:
Also, the distribution would be different for something that returns value type: It would show that operating on `object`s has extra overhead due to boxing.

Reply:
BenchmarkDotNet = instant pit of success from what I've heard.

Reply:
An advantage of the "hop" approach, however, is that it will work with linkers since it has the direct calls to the setters\getters. The same can't be said for some of the other approaches.

Reply:
I agree with that.

Reply:
To be precise:
- It works with the linker if you can reliably identify the full of types involved in the serialization. 
- If you use the strongly typed wrappers, it still requires code generation at runtime (e.g. hidden behind MakeGenericType, etc.)

Reply:
> I think you are seeing a lot of noise from tiered JIT

Yes thanks the numbers do even out after a few runs.

However, if we are measuring cold start we do care about the first run.

Reply:
If you would like to measure cold start, you need to measure it properly. I do not think that the numbers from the first iteration say anything about the cold start characteristics of the different options.

Reply:
Alsothe "MethodInfo getter" or variant of that approach would be useful to remove our Emit usage, which we are tracking in https://github.com/dotnet/runtime/issues/23716.

Reply:
For example, the expression getter should have by far the worst cold start characteristic because of it pulls in Expressions that are big expensive piece of code.

Reply:
Yes definitely the Emit and Expression approaches are out for cold start.

Since today we use Emit I think we likely have two options:
1) For AOT, a code-gen'd delegate approach as shown in the doc's prototype will be fast and work with linkers.
2) For non-AOT a new reflection feature is needed that does not have cold start issues but still has fast steady-state performance.

Reply:
For non-AOT, until the "new reflection feature" is available, I suggest we continue to do what we do today which is use Emit where it is supported and use the slow Reflection.Invoke as a fallback.

Reply:
Ideally we have one approach for both AOT and non-AOT but until we get further along with the "new reflection feature" I'm not sure if that will be feasible.

We also need to determine the priority of linker support, and\or whether the new reflection feature could support the linker in some way.

Reply:
> We also need to determine the priority of linker support

+1. I think there is no shared understanding of it today.

Would it make sense to simplify this to just two scenarios that we focus on?
- The dynamic mode that we have today
- The linker and AOT friendly mode

I understand you can have number of options in between, but I am worried that having many different options will be hard to explain.